src/Libraries/CoreNodes/List.cs

```diff
@@ -525,6 +525,162 @@ public static IDictionary GroupByKey(IList list, IList keys)
+            {
+                var start = 0;
+                for (int i = 1; i < list.Count; i++)
+                {
+                    if (AreItemsSimilar(list[i - 1], list[i], tolerance))
+                    {
+                        currentGroup.Add(list[i]);
+                    }
+                    else
+                    {
+                        result.Add(currentGroup);
+                        idxResult.Add(Enumerable.Range(start, currentGroup.Count).ToList());
+                        currentGroup = new ArrayList { list[i] };
+                        start = i;
+                    }
+                }
+                result.Add(currentGroup);
+                idxResult.Add(Enumerable.Range(start, currentGroup.Count).ToList());
+            }
+            else
+            {
+                var groupedItems = new List<ArrayList>();
+                var itemGroups = new List<object>();
+                var indexGroups = new List<ArrayList>();
+
+                for (int i = 0; i < list.Count; i++)
+                {
+                    var item = list[i];
+                    bool added = false;
+                    for (int j = 0; j < itemGroups.Count; j++)
+                    {
+                        if (AreItemsSimilar(itemGroups[j], item, tolerance))
+                        {
+                            groupedItems[j].Add(item);
+                            indexGroups[j].Add(i);
+                            added = true;
+                            break;
+                        }
+                    }
+
+                    if (!added)
+                    {
+                        itemGroups.Add(item);
+                        groupedItems.Add(new ArrayList { item });
+                        indexGroups.Add(new ArrayList { i });
+                    }
+                }
+
+                foreach (var group in groupedItems)
+                {
+                    result.Add(group);
+                }
+
+                foreach (var group in indexGroups)
+                {
+                    idxResult.Add(group);
+                }
+            }
+
+            return new Dictionary<object, object>
+            {
+                { "groupedValues", result },
+                { "groupedIndices", idxResult }
+            };
+        }
+
+        private static bool AreItemsSimilar(object item1, object item2, int tolerance)
+        {
+            if (item1 is string str1 && item2 is string str2)
+            {
+                if(tolerance > 5)
+                    throw new ArgumentException("Tolerance value for string comparison cannot be greater than 5.", nameof(tolerance));
+                return LevenshteinDistance(str1, str2) <= tolerance;
+            }
+            else if (item1 is IComparable comp1 && item2 is IComparable comp2)
+            {
+                try
+                {
+                    return Math.Abs(Convert.ToDouble(comp1) - Convert.ToDouble(comp2)) <= tolerance;
```

Comment:
Why are you only handling the case where the `IComparables` are convertible to doubles? What if they can also be cast to other types? If they are not convertible to doubles, shouldn't you try doing something like `return  comp1.CompareTo(comp2)`?

Reply:
We need to consider cases with tolerance as well, using `CompareTo ` would not provide us the difference between the two items.

Reply:
Then how about supporting strings as a separate case?

Reply:
we do have a separate case for strings